startup/GafferOSL/oslExamples.py

```diff
@@ -0,0 +1,27 @@
+import Gaffer
+import GafferUI
+import GafferOSL
+
+GafferUI.Examples.registerExample(
+	"Procedurals/OSLImage Node Basics",
```

Comment:
I briefly considered having an "OSL" sub-menu, but "Procedurals" is probably more useful. I'm open to suggestions on better wording.

Reply:
I prefer `OSL` over `Procedurals` mainly as a lot of people are asking for "OSL" examples for all uses so it will make it more discoverable from the top level `Examples` sub-menu. Feels like it's quite high up on their mental search terms. Especially if we also want to add any OSL expression examples over time.

I guess it really depends on wether in the long run we want to be re-enforcing the 'OSL' aspect, rather than the result. It feels like this is OK as for new users, Gaffer employs of OSL in many contexts that other applications don't - so the focus on OSL over say 'Image Processing' and 'Scene Processing' is valuable.  If we did want to stick with the existing pattern of more _what_ than _how_, then it'd probably be better to just add them in to the more general 'contextual' categories - splitting up the `OSLImage` and `OSLObject` ones accordingly as their 'what' is probably quite different in those terms.
 
`Procedurals` _may_ also be a slightly loaded term in this context too. Despite the general definition of the word, conversationally, I've always heard 'procedural' being used for parameterised 'generators'. Ie: textures, or geometry generation or modifiers that make use of some internal generator component.  As the scope of OSL with gaffer is somewhat broader. eg: general image processing rather than synthesis, people may not think to look in there. It'd be good to avoid that ambiguity for the users it does exist for.

As a more general comment as an extension of the above. Looking through the current examples, we seem to have tried to also ensure the name of the example tells you what it does without you needing to knowing the how. With these ones, you have to open and read the backdrops just to know what the results are. What do you think about splitting them up a bit such that we can then give a clue to the what?Otherwise you have to know you want `OSLObject`/`OSLImage` first, which seems more appropriate for the reference docs. eg something generally along the lines of:

 - OSL
      - Mesh Manipulation
      - Image Generation
      - Image Manipulation

or, if we went with the other grouping option:

 - Scene Processing 
     - OSL Mesh Manipulation
 - Rendering
     - OSL Image/Texture Generation
 - Compositing
    - OSL Image Processing
    - OSL Image/Texture Generation

(I just noticed too, we should probably move `Blockers` to be under `Lighting`)

I guess I probably prefer the second one as it isn't making exceptions to our existing patterns.

Reply:
> Procedurals may also be a slightly loaded term in this context too.

This was my first thought. Without additional clarification, "Procedural" alone tends to be used to refer to render-time geometry generators, like Yeti or XGen. In other contexts it tends to be clarified with an additional word, for instance "Procedural texture" or "Procedural shader".

> Scene Processing
      - OSL Mesh Manipulation

This seems like a good way of organising it to me. If we wanted to use "Procedural" rather than "OSL" then that could also work : SceneProcessing->ProceduralMeshManipulation.

Reply:
> This seems like a good way of organising it to me. If we wanted to use "Procedural" rather than "OSL" then that could also work : SceneProcessing->ProceduralMeshManipulation.

 I think we definitely need `OSL` in the menu _somewhere_ if we can. As get the feeling from chats I've had that people may hear "Oh you can do that with OSL", and it helps link the two together when they then attempt the task of "getting on with it".

Reply:
Thanks for the info on the usage of _procedural_.

I'd like to keep the menu nesting shallow, so to me the optimal choices are:

- _Scene Processing_ > _OSL Mesh Manipulation_
- _OSL_ > _Mesh Manipulation_

> ... the focus on OSL over say 'Image Processing' and 'Scene Processing' is valuable. If we did want to stick with the existing pattern of more _what_ than _how_, then it'd probably be better to just add them in to the more general 'contextual' categories...

If the consensus is that _OSL_ is a prominent enough high-level concept/domain, then I can argue that it sticks to this pattern, because the _what_ (or perhaps _where_) would likely include OSL itself, e.g.:

> As a lookdev artist, I would like to **adjust a primvar using OSL.**

The _how_ in this case would be the specific nodes, their connections, and key plugs.

Are we expecting that the user will already know of the domain split? If so, we should be able to safely use _OSL_ as a signpost.

Reply:
> Are we expecting that the user will already know of the domain split? If so, we should be able to safely use _OSL_ as a signpost.

I don't think so. OSL in itself might be well known, but probably only in the context of shader writing. Gaffer is relatively unique in the sense that it uses OSL for a myriad of other things that aren't shading.

As such, I see these examples and the associated learning as being _"Hey, the thing you want to do, you do it using OSL in Gaffer"_.

So I think I'd go with something like the following split of examples, that covers the three main 'whats':

`Scene Processing` > `OSL Mesh Manipulation`
`Rendering` > `OSL Texture Generation`
`Compositing` > `OSL Image Processing`

Reply:
1258dda

I'm sticking with one our "one example, one file" policy, so for OSLImage I settled on "OSL Image Processing". I don't want to start causing confusion by having multiple registries for the same file. If we decide to make an "OSL Texture Generation" example, we can fork it off into its own example, but I'd rather have more robust examples and discussion of what the OSL code is doing.